### PR TITLE
DotNetCore rework

### PR DESCRIPTION
**Pull request by Lexcess**
Minimal changes required to move solution over to .Net core. Some incompatible and unnecessary dependencies removed. Some unsupportable tests removed (exception serialization). This allows the code to run on Windows/Linux/OSX without the old mono runtime.

### CODE REVIEW

**Comment by Lexcess** [comment]
By the way, the builds and tests do pass, I just changed the build system over to our internal one check the Sample code builds under EU (there is one for Linux and one for Windows)

File: src/CSharp/MetadataWebApi/MetadataWebApi.Tests/IMetadataApiExtensionsTests.cs [hunk 1]
```diff
@@ -25,29 +24,5 @@ public static class IMetadataApiExtensionsTests
             // Act and Assert
             await Assert.ThrowsAsync<ArgumentNullException>("value", () => value.GetDownloadUriAsync(fileName, fileHash));
         }
-
```
Comment (johndarv) [inline src/CSharp/MetadataWebApi/MetadataWebApi.Tests/IMetadataApiExtensionsTests.cs]: @Lexcess Didn't consider deleting the other test while you were at it? Not sure how much value checks for ArgumentNullExceptions bring.

File: src/CSharp/MetadataWebApi/MetadataWebApi.Tests/MetadataApiFactoryTests.cs [hunk 2]
```diff
@@ -5,22 +5,39 @@
...
+
+        private IDictionary<string,string> inMemoryConfigEmpty = new Dictionary<string,string>
+        {
+            {"appSettings:token", ""},
+            {"appSettings:downloadRootPath", ""},
+            {"appSettings:validateDownloads", ""},
+            {"appSettings:serviceUri", ""}
+        };
+
+    [Fact]
```
Comment (johndarv) [inline src/CSharp/MetadataWebApi/MetadataWebApi.Tests/MetadataApiFactoryTests.cs]: @Lexcess Indentation error :)

File: src/CSharp/MetadataWebApi/MetadataWebApi.Tests/MetadataApiFactoryTests.cs [hunk 3]
```diff
@@ -90,39 +103,33 @@ public static void MetadataApiFactory_CreateMetadataApi_Creates_Instance_If_Inva
...
-        public static void MetadataApiFactory_CreateMetadataApi_Throws_If_No_Token_Configured()
+        public void MetadataApiFactory_CreateMetadataApi_Throws_If_No_Token_Configured()
         {
             // Arrange
-            Mock<MetadataApiFactory> mock = new Mock<MetadataApiFactory>();
-
-            mock.CallBase = true;
-            mock.Setup((p) => p.GetConfigSetting("Token")).Returns(string.Empty);
+            var configuration = new ConfigurationBuilder()
+                .AddInMemoryCollection(inMemoryConfigEmpty)
```
Comment (johndarv) [inline src/CSharp/MetadataWebApi/MetadataWebApi.Tests/MetadataApiFactoryTests.cs]: @Lexcess this.inMemoryConfigEmpty or _inMemoryConfigEmpty?

**Comment by sdcondon** [comment]
Don't forget to update the pre-reqs documented in the [MetadataWebApi readme](/src/CSharp/MetadataWebApi/README.md).

File: src/CSharp/MetadataWebApi/MetadataWebApi/MetadataApiFactory.cs [hunk 4]
```diff
@@ -27,15 +33,15 @@ public class MetadataApiFactory : IMetadataApiFactory
         public virtual IMetadataApi CreateMetadataApi()
         {
             // Get the token to use to connect to the QAS Electronic Updates Metadata REST API
-            string token = GetConfigSetting("Token");
+            string token = GetConfigSetting("token");
 
             if (string.IsNullOrEmpty(token))
             {
-                throw new ConfigurationErrorsException("The Electronic Updates authentication token has not been configured.");
+                throw new NullReferenceException("The Electronic Updates authentication token has not been configured.");
```
Comment (johndarv) [inline src/CSharp/MetadataWebApi/MetadataWebApi/MetadataApiFactory.cs]: @Lexcess Don't think we want to be throwing NullReferenceExceptions manually do we?

File: src/CSharp/MetadataWebApi/MetadataWebApi.Tests/IMetadataApiExtensionsTests.cs (hunk 1 above)
Comment (Lexcess) [inline src/CSharp/MetadataWebApi/MetadataWebApi.Tests/IMetadataApiExtensionsTests.cs]: @johndarv I was a little torn... there are quite a few tests and behaviours of questionable value, but this one at least was compatible with .Net Core. In any case I'm trying to change the minimum here but feel free to submit your own PR afterwards! :)

File: src/CSharp/MetadataWebApi/MetadataWebApi/MetadataApiFactory.cs (hunk 4 above)
Comment (Lexcess) [inline src/CSharp/MetadataWebApi/MetadataWebApi/MetadataApiFactory.cs]: @johndarv Which exception would you suggest? ArgumentException wouldn't fit. The whole Configuration namespace has been removed (including the original exception)

File: src/CSharp/MetadataWebApi/MetadataWebApi.Tests/MetadataApiFactoryTests.cs (hunk 3 above)
Comment (Lexcess) [inline src/CSharp/MetadataWebApi/MetadataWebApi.Tests/MetadataApiFactoryTests.cs]: I prefer without but the code here is inconsistent. _ was originally used by MS code to minimize the possibility of collisions with usercode.

File: src/CSharp/MetadataWebApi/MetadataWebApi.Tests/MetadataApiFactoryTests.cs (hunk 2 above)
Comment (Lexcess) [inline src/CSharp/MetadataWebApi/MetadataWebApi.Tests/MetadataApiFactoryTests.cs]: Fair enough

File: src/CSharp/MetadataWebApi/MetadataWebApi/MetadataWebApi.csproj [hunk 5]
```diff
@@ -1,118 +1,23 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (HectorHS) [inline src/CSharp/MetadataWebApi/MetadataWebApi/MetadataWebApi.csproj]: needed to add xmlns="http://schemas.microsoft.com/developer/msbuild/2003" for vs 2015 to load the solution whithout complaining

File: src/CSharp/MetadataWebApi/MetadataWebApi/MetadataApiFactory.cs (hunk 4 above)
Comment (johndarv) [inline src/CSharp/MetadataWebApi/MetadataWebApi/MetadataApiFactory.cs]: @Lexcess InvalidOperationException?

Comment (johndarv) [inline src/CSharp/MetadataWebApi/MetadataWebApi/MetadataApiFactory.cs]: @Lexcess ConfigurationErrorsException maybe?

File: src/CSharp/MetadataWebApi/MetadataWebApi/MetadataWebApi.csproj [hunk 6]
```diff
@@ -1,118 +1,23 @@
...
-    <None Include="packages.config" />
-  </ItemGroup>
-  <ItemGroup>
-    <CodeAnalysisDictionary Include="..\FxCopDictionary.xml">
-      <Link>FxCopDictionary.xml</Link>
-    </CodeAnalysisDictionary>
+    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="1.1.2" />
+    <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
+    <PackageReference Include="system.runtime.serialization.xml" Version="4.3.0" />
+    <PackageReference Include="system.xml.xmlserializer" Version="4.3.*" />
```
Comment (HectorHS) [inline src/CSharp/MetadataWebApi/MetadataWebApi/MetadataWebApi.csproj]: getting an error when opening in vs 2015: "The attribute "Version" in element <PackageReference> is unrecognized."

removing the versions entirely i get a "one-way upgrade" pop up 
"Visual Studio will automatically make functional changes to the following projects in order to open them. You will not be able to open these projects in the version of Visual Studio in which they were originally created."

pressing ok still does not load the project properly

File: src/CSharp/MetadataWebApi/MetadataWebApi/MetadataApiFactory.cs (hunk 4 above)
Comment (Lexcess) [inline src/CSharp/MetadataWebApi/MetadataWebApi/MetadataApiFactory.cs]: As discussed ConfigurationErrorsException is no longer available. I felt that InvalidOperation was a little misleading but it is probably closer than NullReference, I'll change to that.

File: src/CSharp/MetadataWebApi/MetadataWebApi/MetadataWebApi.csproj (hunk 5 above)
Comment (Lexcess) [inline src/CSharp/MetadataWebApi/MetadataWebApi/MetadataWebApi.csproj]: Yeah, I'll update the docs to say that either Visual Studio 2017, Visual Studio Code or dotnet command line is now required.

File: src/CSharp/MetadataWebApi/MetadataWebApi/MetadataWebApi.csproj (hunk 6 above)
Comment (Lexcess) [inline src/CSharp/MetadataWebApi/MetadataWebApi/MetadataWebApi.csproj]: Related to above, it's a VS2017 format but will work with Code or dotnet cmd as well.

**Comment by Lexcess** [comment]
@sdcondon Updated the Readme with new [instructions](https://github.com/experiandataquality/electronicupdates/pull/44/files#diff-e9ef2c48a32ed0200900032911b23a22). 

@HectorHS You can follow the build link for EU-Samples-CI and download the artifacts if you just want to quickly test the actual application

@Johndarv made all the changes

**Comment by Lexcess** [comment]
Thanks for all the effort everyone. We've just had a QA Pass and everything seems green so merging this in now.

**Pull request merged**